src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,501 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+     */
+    ss::future<> progress_conditions(op_context& octx) {
+        ss::promise<> timeout_pr, abort_pr;
+
+        auto abort_sub_opt = _as.subscribe(
+          [&abort_pr]() noexcept { abort_pr.set_value(); });
+
+        if (!abort_sub_opt) {
+            abort_pr.set_value();
+        }
+
+        ss::timer<model::timeout_clock> s{
+          [&timeout_pr] { timeout_pr.set_value(); }};
+
+        if (octx.deadline) {
+            s.arm(octx.deadline.value());
+        }
+
+        co_await ss::when_any(
+          ignore_exceptions(_has_data.wait()),
+          ignore_exceptions(timeout_pr.get_future()),
+          ignore_exceptions(abort_pr.get_future()));
+    }
+
+    /*
+     * `start_shard_fetch_worker` executes on the coordinator shard. It builds
+     * the `shard_local_fetch_context` struct needed to start the
+     * shard_fetch_worker then makes a cross shard call to the shard fetch
+     * worker if needed.
+     *
+     * It then waits until the shard_fetch_worker completes
+     * and then notifies the fetch coordinator.
+     */
+    ss::future<> start_shard_fetch_worker(
+      ss::shard_id shard,
+      op_context& octx,
+      shard_fetch fetch,
+      size_t min_fetch_bytes) {
+        // if over budget skip the fetch.
+        if (octx.bytes_left <= 0) {
+            co_return;
+        }
+        // no requests for this shard, do nothing
+        if (fetch.empty()) {
+            co_return;
+        }
+
+        const bool foreign_read = shard != ss::this_shard_id();
+
+        auto results = co_await octx.rctx.partition_manager().invoke_on(
+          shard,
+          octx.ssg,
+          [this,
+           min_fetch_bytes,
+           foreign_read,
+           configs = fetch.requests,
+           &octx](cluster::partition_manager& mgr) mutable {
+              // Although this and octx are captured by reference across shards
+              // it is safe since they are not modified by any shard for the
+              // duration of the capture and they are protected from deletion
+              // for the duration of the capture by a gate. Both are used
+              // immediately on the foreign shard to access data local to that
+              // shard. This is meant to help avoiding unintended cross shard
+              // access.
+              return shard_fetch_worker(shard_local_fetch_context{
+                .foreign_read = foreign_read,
+                .bytes_left = octx.bytes_left,
+                .min_bytes = min_fetch_bytes,
+                .deadline = octx.deadline,
+                .requests = std::move(configs),
+                .srv = octx.rctx.server().local(),
+                .mgr = mgr,
+                .as = _as.local(),
+              });
+          });
+
+        fill_fetch_responses(
+          octx,
+          std::move(results.read_results),
+          fetch.responses,
+          fetch.start_time);
+
+        // Remove errored fetches from fetch.requests and fetch.responses.
```

Comment:
Why do this? Do we have something equivalent in the old code? Why the sideband passing of errored partitions then this sorting & shifting thing versus indicatin the error in the results.read_results object?

Reply:
Originally I wasn't sure if the fetch request returns as soon as a sub-fetch fails. This therefore was to remove any sub-fetches that had errors so we wouldn't waste time re-querying them. Now that I'm certain a fetch will return immediately it does there isn't a need for this. Will remove.